src/python/grpcio/grpc/_cython/_cygrpc/server.pyx.pxi

```diff
@@ -17,6 +17,36 @@ cimport cpython
 import time
 
 
+cdef grpc_ssl_certificate_config_reload_status _server_cert_config_fetcher_wrapper(
+        void* user_data, grpc_ssl_server_certificate_config **config) with gil:
+  # We are not catching any exception here, because cython will
+  # happily catch and ignore it, and will log for us, and also the
+  # core lib will continue as if the cert has not changed, which is a
+  # reasonable behavior.
```

Comment:
thanks for the tidbit on PHP; it is funny and definitely not something I would subscribe to :)
the exceptions i'm referring to are mainly from the user's fetcher callback. It would be totally advocating for crashing; please let me know how to do that, i.e., instruct cython to not catch and ignore exceptions here.

Another option is to manually catch exceptions here and return `GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_ERROR`; however, the outcome would not be materially different, since the C-core will again just log the error and continue as if the certificate has not changed.

Reply:
That second option sounds like the right thing to do - if the Core calls us (and it does) and it allows "the application messed up!" as an answer (and it does) then we should tell it that the application messed up in cases in which the application messed up.

That the Core decides to keep on serving is a decision with which I disagree, but... it's out of scope of this change to fuss over it.

Reply:
will do. something like this?
```
else:
  try:
    cert_config_wrapper = user_cb()
  except:
    return GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_ERROR
  if cert_config_wrapper is None:
    return GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_UNCHANGED
  # what if the cert_config_wrapper is not of the right type? i.e.,
  # an instance of grpc.ServerCertificateConfig?
```
what kind of logging should we / can we do here? none of the `pyx.pxi` files does logging, so i don't have an example to follow :)

Reply:
`except Exception:` rather than `except:`.

I think we should just do plain old ordinary Python logging: `logging.exception(<string message>)`.

Reply:
i've tried `logging.error("blah blah")` but i can't see it in stdout/error of the test, even though i can see other error messages like:
```
stderr:
D1114 00:21:02.601111581   23569 security_connector.cc:641]  Using new server certificate config (0x7f08e000d720).
D1114 00:21:02.646434865   23570 security_connector.cc:688]  No change in SSL server credentials.
Traceback (most recent call last):
  File "/home/cauthu/my-grpc-fork/src/python/grpcio_tests/tests/unit/_server_ssl_cert_config_test.py", line 88, in __call__
    'just an error for fun, should not affect the test')
ValueError: just an error for fun, should not affect the test
Exception ValueError: ValueError('just an error for fun, should not affect the test',) in 'grpc._cython.cygrpc._server_cert_config_fetcher_wrapper' ignored
```

Reply:
tl;dr: done.

so it seems like the logging does work when i run the a server outside of the testing environment: this is in `server.pyx.pxi`:
```
except Exception:
    logging.exception('Error fetching certificate config')
    return GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_FAIL
...
elif not isinstance(cert_config_wrapper, ServerCertificateConfig):
    logging.error('Error fetching certificate config: certificate '
                  'config must be of type grpc.ServerCertificateConfig')
```
* make the config fetcher always raise an exception:
```
$ python foo.py server 1>/dev/null
ERROR:root:Error fetching certificate config
Traceback (most recent call last):
  File "src/python/grpcio/grpc/_cython/_cygrpc/server.pyx.pxi", line 35, in grpc._cython.cygrpc._server_cert_config_fetcher_wrapper
  File "foo.py", line 57, in __call__
    raise Exception('ALWAYS RAISE')
Exception: ALWAYS RAISE
E1114 02:10:39.476019806   10443 security_connector.cc:694]  Failed fetching new server credentials, continuing to use previously-loaded credentials.
^CTraceback (most recent call last):
  File "foo.py", line 110, in <module>
    time.sleep(100)
KeyboardInterrupt
$
```
* make the config fetcher always return a `1`:
```
$ python foo.py server 1>/dev/null
ERROR:root:Error fetching certificate config: certificate config must be of type grpc.ServerCertificateConfig
E1114 02:11:23.706952551   10474 security_connector.cc:694]  Failed fetching new server credentials, continuing to use previously-loaded credentials.
^CTraceback (most recent call last):
  File "foo.py", line 111, in <module>
    time.sleep(100)
KeyboardInterrupt
$
```

Reply:
ok so it's probably error on my part that the logging messages did not show up in stderr during the test while they did when i ran programs outside the test: it might have been because i needed to remove `py27` (and then having to recreate it with `virtualenv py27`) before running the tests; otherwise the tests wouldn't pick up the change that i made to add the `logging.exception()` call.

now that i've done that--ensure the tests use the updated code--i can see the expected error messages, e.g.:
```
D1114 03:00:13.398917976    5456 security_connector.cc:688]  No change in SSL server credentials.
ERROR:root:Error fetching certificate config: certificate config must be of type grpc.ServerCertificateConfig, not list
E1114 03:00:13.428699928    5456 security_connector.cc:694]  Failed fetching new server credentials, continuing to use previously-loaded credentials.


Unexpected successes: []
Test failure
FAILED: py27.test.unit._server_ssl_cert_config_test.ServerSSLCertReloadTestWithClientAuth [ret=1, pid=5412, time=1.4sec]
```